gh-99337: Fix compile errors with gcc 12 on macOS

ronaldoussoren opened the pull request:
When compiling with GCC 12 on a macOS 12 system I get three groups of errors:

1. In pylifecycle.c the compile rejects _Pragma within a declaration
2. posixmodule.c was missing a number of ..._RUNTIME macros for non-clang on macOS
3. _ctypes assumed that __builtin_available is always present on macOS

<!-- gh-issue-number: gh-99337 -->
* Issue: gh-99337
<!-- /gh-issue-number -->

ronaldoussoren committed:
GCC 12 complains about _Pragma inside a decleration

ronaldoussoren committed:
Fix a number of compile errors with GCC-12 on macOS

ronaldoussoren committed:
Centralize definition of HAVE_BUILTIN_AVAILABLE in the ctypes extension

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

vstinner commented on Modules/_ctypes/malloc_closure.c:
```diff
@@ -96,7 +97,11 @@ void Py_ffi_closure_free(void *p)
+# ifdef HAVE_BUILTIN_AVAILABLE
     if (__builtin_available(macos 10.15, ios 13, watchos 6, tvos 13, *)) {
+#  else
+    if (&ffi_closure_free != NULL) {
```
Is it ``&ffi_closure_free`` or ``ffi_closure_free`` which can be NULL?

ronaldoussoren replied:
Testing for ffi_closure_free matches the pattern in other files, and in particular posixmodule.c where we have a longer history of weak linking (although the truly ancient parts of that were removed when adding support for arm64)

ronaldoussoren replied:
Note that I expect that this will be a no-op test for most users, it can only be false when targeting an older version of macOS.

vstinner commented on Python/pylifecycle.c:
```diff
@@ -83,6 +83,8 @@ int _Py_UnhandledKeyboardInterrupt = 0;
  * found without any external information. This is meant to ease access to the
  * interpreter state for various runtime debugging tools, but is *not* an
  * officially supported feature */
+_Py_COMP_DIAG_PUSH
```
You should move and update ``/* Suppress deprecation warning for PyBytesObject.ob_shash */`` comment, no?

ronaldoussoren replied:
Agreed.

ronaldoussoren committed:
Adjust test for ffi_closure_free weak symbol

The new form matches the way we test for other weak
symbols (for example in posixmodule.c)

ronaldoussoren committed:
Move the comment about suppressing deprecation warnings

I earlier moved the pragma, but not the comment that explains
why it is necessary.

vstinner reviewed (approved):
LGTM.

Oh ok, now with the fixed test ``if (ffi_closure_free != NULL) {``, I get it :-)
